File: examples/webhooks/fastapi/pre_call.py
"""
ElevenLabs pre-call webhook — FastAPI implementation.

Performs a CRM lookup and conversation history fetch before the call connects,
returning dynamic variables and a personalized first message.
"""

from __future__ import annotations

import asyncio
import logging
import os
from datetime import datetime, timezone
from typing import Any

import httpx
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

CRM_API_KEY: str = os.environ["CRM_API_KEY"]
ELEVENLABS_API_KEY: str = os.environ["ELEVENLABS_API_KEY"]
ELEVENLABS_AGENT_ID: str = os.environ["ELEVENLABS_AGENT_ID"]

CRM_BASE_URL: str = os.environ.get("CRM_BASE_URL", "https://api.yourcrm.com/v1")
ELEVENLABS_BASE_URL: str = "https://api.elevenlabs.io/v1"

REQUEST_TIMEOUT_SECS: float = 3.5
HANDLER_TIMEOUT_SECS: float = 4.5

# ---------------------------------------------------------------------------
# Models
# ---------------------------------------------------------------------------


class CrmContact(BaseModel):
    id: str
    first_name: str
    last_name: str
    email: str
    company: str
    account_status: str  # active | churned | trial | prospect
    lifetime_value: float
    last_interaction: str
    notes: str = ""


class ConversationAnalysis(BaseModel):
    call_successful: str | None = None
    transcript_summary: str | None = None


class ElevenLabsConversation(BaseModel):
    conversation_id: str
    agent_id: str
    status: str
    metadata: dict[str, str] = Field(default_factory=dict)
    analysis: ConversationAnalysis | None = None
    start_time_unix_secs: int = 0


class PreCallEvent(BaseModel):
    call_sid: str
    caller_id: str
    called_number: str
    call_type: str  # phone | web
    agent_id: str


class PreCallWebhookPayload(BaseModel):
    type: str
    event: PreCallEvent


class AgentOverride(BaseModel):
    first_message: str | None = None


class ConversationConfigOverride(BaseModel):
    agent: AgentOverride | None = None


class PreCallWebhookResponse(BaseModel):
    dynamic_variables: dict[str, str] = Field(default_factory=dict)
    conversation_config_override: ConversationConfigOverride | None = None


# ---------------------------------------------------------------------------
# HTTP helpers
# ---------------------------------------------------------------------------


async def lookup_crm_contact(
    client: httpx.AsyncClient,
    phone_number: str,
) -> CrmContact | None:
    """Look up a contact by phone number in the CRM."""
    resp = await client.get(
        f"{CRM_BASE_URL}/contacts/search",
        params={"phone": phone_number},
        headers={
            "Authorization": f"Bearer {CRM_API_KEY}",
            "Content-Type": "application/json",
        },
        timeout=REQUEST_TIMEOUT_SECS,
    )
    if resp.status_code == 404:
        return None
    resp.raise_for_status()

    contacts = resp.json().get("contacts", [])
    return CrmContact(**contacts[0]) if contacts else None


async def get_conversation_history(
    client: httpx.AsyncClient,
    caller_id: str,
) -> list[ElevenLabsConversation]:
    """Fetch recent conversations for this agent and filter by caller."""
    resp = await client.get(
        f"{ELEVENLABS_BASE_URL}/convai/conversations",
        params={"agent_id": ELEVENLABS_AGENT_ID, "page_size": "5"},
        headers={"xi-api-key": ELEVENLABS_API_KEY},
        timeout=REQUEST_TIMEOUT_SECS,
    )
    resp.raise_for_status()

    conversations = [
        ElevenLabsConversation(**c)
        for c in resp.json().get("conversations", [])
    ]
    return [c for c in conversations if c.metadata.get("caller_id") == caller_id]


# ---------------------------------------------------------------------------
# Context builders
# ---------------------------------------------------------------------------


def build_caller_context(
    contact: CrmContact | None,
    history: list[ElevenLabsConversation],
) -> str:
    parts: list[str] = []

    if contact:
        parts.append(f"Caller: {contact.first_name} {contact.last_name}")
        parts.append(f"Company: {contact.company}")
        parts.append(f"Account status: {contact.account_status}")
        if contact.notes:
            parts.append(f"CRM notes: {contact.notes}")

    if history:
        parts.append(f"Previous calls: {len(history)}")
        last_call = history[0]
        if last_call.analysis and last_call.analysis.transcript_summary:
            parts.append(
                f"Last call summary: {last_call.analysis.transcript_summary}"
            )
        last_date = datetime.fromtimestamp(
            last_call.start_time_unix_secs, tz=timezone.utc
        ).strftime("%b %d, %Y")
        parts.append(f"Last call date: {last_date}")

    return "\n".join(parts) if parts else "No prior contact information available."


def build_first_message(contact: CrmContact | None) -> str:
    if contact:
        return (
            f"Hi {contact.first_name}, thanks for calling! "
            "How can I help you today?"
        )
    return "Thanks for calling! How can I help you today?"


def fallback_response() -> PreCallWebhookResponse:
    return PreCallWebhookResponse(
        dynamic_variables={
            "caller_context": "No prior contact information available.",
            "caller_name": "Unknown",
            "account_status": "unknown",
            "previous_call_count": "0",
        },
    )


# ---------------------------------------------------------------------------
# Router
# ---------------------------------------------------------------------------

router = APIRouter()


@router.post(
    "/webhooks/elevenlabs/pre-call",
    response_model=PreCallWebhookResponse,
)
async def pre_call_webhook(request: Request) -> PreCallWebhookResponse:
    try:
        body = await request.json()
        payload = PreCallWebhookPayload(**body)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=f"Invalid payload: {exc}")

    caller_id = payload.event.caller_id

    async def _run() -> PreCallWebhookResponse:
        async with httpx.AsyncClient() as client:
            crm_task = asyncio.create_task(
                lookup_crm_contact(client, caller_id)
            )
            history_task = asyncio.create_task(
                get_conversation_history(client, caller_id)
            )

            results = await asyncio.gather(
                crm_task, history_task, return_exceptions=True
            )

        contact: CrmContact | None = None
        history: list[ElevenLabsConversation] = []

        if isinstance(results[0], BaseException):
            logger.error("CRM lookup failed: %s", results[0])
        else:
            contact = results[0]

        if isinstance(results[1], BaseException):
            logger.error("History fetch failed: %s", results[1])
        else:
            history = results[1]

        caller_context = build_caller_context(contact, history)
        first_message = build_first_message(contact)

        return PreCallWebhookResponse(
            dynamic_variables={
                "caller_context": caller_context,
                "caller_name": (
                    f"{contact.first_name} {contact.last_name}"
                    if contact
                    else "Unknown"
                ),
                "account_status": contact.account_status if contact else "unknown",
                "previous_call_count": str(len(history)),
            },
            conversation_config_override=ConversationConfigOverride(
                agent=AgentOverride(first_message=first_message),
            ),
        )

    try:
        return await asyncio.wait_for(_run(), timeout=HANDLER_TIMEOUT_SECS)
    except asyncio.TimeoutError:
        logger.warning("Pre-call webhook hit hard timeout, returning fallback")
        return fallback_response()
    except Exception:
        logger.exception("Unhandled error in pre-call webhook")
        return fallback_response()

File: examples/webhooks/express/pre-call.ts
import { Request, Response, Router } from "express";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface CrmContact {
  id: string;
  first_name: string;
  last_name: string;
  email: string;
  company: string;
  account_status: "active" | "churned" | "trial" | "prospect";
  lifetime_value: number;
  last_interaction: string;
  notes: string;
}

interface ElevenLabsConversation {
  conversation_id: string;
  agent_id: string;
  status: string;
  metadata: Record<string, string>;
  analysis?: {
    call_successful?: string;
    transcript_summary?: string;
  };
  start_time_unix_secs: number;
}

interface ConversationListResponse {
  conversations: ElevenLabsConversation[];
  has_more: boolean;
}

interface PreCallWebhookPayload {
  type: "pre_call";
  event: {
    call_sid: string;
    caller_id: string;
    called_number: string;
    call_type: "phone" | "web";
    agent_id: string;
  };
}

interface PreCallWebhookResponse {
  dynamic_variables?: Record<string, string>;
  conversation_config_override?: {
    agent?: {
      first_message?: string;
      prompt?: {
        prompt?: string;
      };
    };
  };
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const CRM_API_KEY = process.env.CRM_API_KEY!;
const ELEVENLABS_API_KEY = process.env.ELEVENLABS_API_KEY!;
const ELEVENLABS_AGENT_ID = process.env.ELEVENLABS_AGENT_ID!;

const CRM_BASE_URL = process.env.CRM_BASE_URL ?? "https://api.yourcrm.com/v1";
const ELEVENLABS_BASE_URL = "https://api.elevenlabs.io/v1";

const REQUEST_TIMEOUT_MS = 3_500;
const HANDLER_TIMEOUT_MS = 4_500;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function timedFetch(
  url: string,
  init: RequestInit & { timeout?: number } = {},
): Promise<globalThis.Response> {
  const controller = new AbortController();
  const timeout = init.timeout ?? REQUEST_TIMEOUT_MS;
  const timer = setTimeout(() => controller.abort(), timeout);

  return fetch(url, {
    ...init,
    signal: controller.signal,
  }).finally(() => clearTimeout(timer));
}

async function lookupCrmContact(
  phoneNumber: string,
): Promise<CrmContact | null> {
  const res = await timedFetch(
    `${CRM_BASE_URL}/contacts/search?phone=${encodeURIComponent(phoneNumber)}`,
    {
      headers: {
        Authorization: `Bearer ${CRM_API_KEY}`,
        "Content-Type": "application/json",
      },
    },
  );

  if (!res.ok) {
    if (res.status === 404) return null;
    throw new Error(`CRM lookup failed: ${res.status} ${res.statusText}`);
  }

  const data = (await res.json()) as { contacts: CrmContact[] };
  return data.contacts.length > 0 ? data.contacts[0] : null;
}

async function getConversationHistory(
  callerId: string,
): Promise<ElevenLabsConversation[]> {
  const url = new URL(`${ELEVENLABS_BASE_URL}/convai/conversations`);
  url.searchParams.set("agent_id", ELEVENLABS_AGENT_ID);
  url.searchParams.set("page_size", "5");

  const res = await timedFetch(url.toString(), {
    headers: {
      "xi-api-key": ELEVENLABS_API_KEY,
    },
  });

  if (!res.ok) {
    throw new Error(
      `ElevenLabs history fetch failed: ${res.status} ${res.statusText}`,
    );
  }

  const data = (await res.json()) as ConversationListResponse;
  return data.conversations.filter(
    (c) => c.metadata?.caller_id === callerId,
  );
}

function buildCallerContext(
  contact: CrmContact | null,
  history: ElevenLabsConversation[],
): string {
  const parts: string[] = [];

  if (contact) {
    parts.push(
      `Caller: ${contact.first_name} ${contact.last_name}`,
      `Company: ${contact.company}`,
      `Account status: ${contact.account_status}`,
    );
    if (contact.notes) {
      parts.push(`CRM notes: ${contact.notes}`);
    }
  }

  if (history.length > 0) {
    parts.push(`Previous calls: ${history.length}`);
    const lastCall = history[0];
    if (lastCall.analysis?.transcript_summary) {
      parts.push(`Last call summary: ${lastCall.analysis.transcript_summary}`);
    }
    const lastDate = new Date(
      lastCall.start_time_unix_secs * 1000,
    ).toLocaleDateString("en-US", {
      month: "short",
      day: "numeric",
      year: "numeric",
    });
    parts.push(`Last call date: ${lastDate}`);
  }

  return parts.length > 0
    ? parts.join("\n")
    : "No prior contact information available.";
}

function buildFirstMessage(contact: CrmContact | null): string {
  if (contact) {
    return `Hi ${contact.first_name}, thanks for calling! How can I help you today?`;
  }
  return "Thanks for calling! How can I help you today?";
}

function fallbackResponse(): PreCallWebhookResponse {
  return {
    dynamic_variables: {
      caller_context: "No prior contact information available.",
      caller_name: "Unknown",
      account_status: "unknown",
      previous_call_count: "0",
    },
  };
}

// ---------------------------------------------------------------------------
// Router
// ---------------------------------------------------------------------------

const router = Router();

router.post(
  "/webhooks/elevenlabs/pre-call",
  async (req: Request, res: Response): Promise<void> => {
    const run = async (): Promise<PreCallWebhookResponse> => {
      const payload = req.body as PreCallWebhookPayload;
      const { caller_id: callerId } = payload.event;

      const [contactResult, historyResult] = await Promise.allSettled([
        lookupCrmContact(callerId),
        getConversationHistory(callerId),
      ]);

      const contact =
        contactResult.status === "fulfilled" ? contactResult.value : null;
      const history =
        historyResult.status === "fulfilled" ? historyResult.value : [];

      if (contactResult.status === "rejected") {
        console.error("CRM lookup failed:", contactResult.reason);
      }
      if (historyResult.status === "rejected") {
        console.error("History fetch failed:", historyResult.reason);
      }

      const callerContext = buildCallerContext(contact, history);
      const firstMessage = buildFirstMessage(contact);

      return {
        dynamic_variables: {
          caller_context: callerContext,
          caller_name: contact
            ? `${contact.first_name} ${contact.last_name}`
            : "Unknown",
          account_status: contact?.account_status ?? "unknown",
          previous_call_count: String(history.length),
        },
        conversation_config_override: {
          agent: {
            first_message: firstMessage,
          },
        },
      };
    };

    const hardTimeout = new Promise<PreCallWebhookResponse>((resolve) => {
      setTimeout(() => {
        console.warn("Pre-call webhook hit hard timeout, returning fallback");
        resolve(fallbackResponse());
      }, HANDLER_TIMEOUT_MS);
    });

    try {
      const response = await Promise.race([run(), hardTimeout]);
      res.json(response);
    } catch (error) {
      console.error("Unhandled error in pre-call webhook:", error);
      res.json(fallbackResponse());
    }
  },
);

export default router;
